### ISSUE TITLE
Native fused linear_logp backend for Ampere

### ISSUE THREAD

**Issue by QingYUDC** [RL-Align/RL-Kernel#439]
## Motivation

The current `linear_logp` fast path is optimized for SM90/Hopper, while A100/SM80 currently dispatches to the Triton implementation.

The RL-Kernel roadmap mentions an SM80 `cp.async` / `mma.sync` fallback for fused linear LogProb. This RFC proposes implementing the missing native SM80 path.

I ran an initial characterization on **2× NVIDIA A100 80GB PCIe (SM80)** using BF16 with:

- hidden size: `4096`
- vocab size: `128256`
- token count: `128–4096`
- FP32-upcast `F.linear + log_softmax` as the numerical reference

On A100, `linear_logp` currently resolves to:

```text
TritonLinearLogpOp
```

For `N = 4096`, the current measurements are:

| Backend | Forward latency | Peak activation |
|---|---:|---:|
| PyTorch / cuBLAS materialized path | ~25.0 ms | ~5.0 GB |
| Triton `linear_logp` | ~82.1 ms | ~0 MB |

The Triton implementation preserves the main memory advantage of fused `linear_logp`, but there is still substantial room to improve latency on Ampere.

The goal is therefore **not simply to add A100 support**, since the existing Triton path already works correctly.

The goal is to provide a native SM80 CUDA backend that combines:

1. no `[N, V]` logits materialization;
2. improved performance on Ampere GPUs;
3. FP32 accumulation and online-softmax numerical behavior;
4. a more explicitly controlled execution path that can support future batch-invariant / train-rollout consistency work.

---

## Proposed scope

The initial implementation will focus on a native SM80 BF16 `linear_logp` forward path.

Target platform:

- NVIDIA A100 / compute capability 8.0
- BF16
- single GPU
- forward path
- FP32 accumulation
- online softmax
- no full logits materialization

The implementation can make use of Ampere-native execution primitives such as:

```text
Global Memory
     |
     | cp.async
     v
Shared Memory
     |
     | ldmatrix
     v
mma.sync
     |
     v
FP32 Accumulation
     |
     v
Online Softmax
     |
     v
Selected LogProb
```

The exact tile sizes, pipeline stages, and reduction strategy will be determined through profiling and benchmark results.

---

### 1. Native SM80 forward kernel

Add a native CUDA implementation for representative RL/LLM workloads.

Initial characterization shape:

```text
H = 4096
V = 128256
dtype = BF16
N = 128 ... 4096
```

The kernel will:

- stream LM-head weight tiles;
- use Tensor Core MMA for projection;
- accumulate in FP32;
- maintain online max / sum-exp statistics;
- track the target-token logit;
- directly produce selected-token LogProb.

The full `[N, V]` logits tensor will not be materialized.

---

### 2. Ampere-specific optimization

After the basic implementation is correct, the SM80 path can be optimized using:

- `cp.async` global-to-shared prefetch;
- shared-memory double buffering;
- `ldmatrix`;
- BF16 `mma.sync`;
- token/vocabulary tile tuning;
- register-pressure tuning;
- pipeline-stage tuning.

The objective is to reduce the current gap between the Triton fused implementation and the highly optimized cuBLAS projection path while retaining the memory advantages of fused `linear_logp`.

---

### 3. Kernel registry integration

The new backend can be integrated into the existing `KernelRegistry`.

Conceptually:

```text
SM90
  -> existing native SM90 backend

SM80 + supported configuration
  -> native SM80 backend

other supported CUDA configurations
  -> Triton backend

fallback
  -> existing reference path
```

The existing SM90, Triton, and PyTorch implementations will remain available.

---

## Validation

The SM80 implementation will be compared against the existing FP32-upcast reference implementation.

Representative test matrix:

```text
N = 128
    256
    512
    1024
    2048
    4096

H = 4096
V = 128256
dtype = BF16
```

Validation will include:

- selected-token LogProb correctness;
- max absolute error;
- mean absolute error;
- different target-token positions;
- different token counts;
- runtime backend selection;
- fallback behavior for unsupported configurations.

The implementation will aim to maintain numerical accuracy comparable to the current fused implementation.

---

## Performance evaluation

The benchmark will compare:

1. PyTorch / cuBLAS materialized reference;
2. current Triton `linear_logp`;
3. native SM80 `linear_logp`.

Metrics:

- forward latency;
- tokens/s;
- peak activation memory;
- numerical error.

The main performance objective is to substantially improve over the current Triton path while preserving the no-logits-materialization property.

For reference, the current A100 baseline at:

```text
H = 4096
V = 128256
N = 4096
BF16
```

is approximately:

```text
PyTorch / cuBLAS:   ~25 ms
Triton fused:       ~82 ms
```

The native SM80 implementation will be evaluated across the full token-count range rather than against a single fixed latency threshold.

Where profiling access is available, the evaluation can additionally include:

- Tensor Core utilization;
- DRAM / L2 throughput;
- achieved occupancy;
- register usage;
- shared-memory usage;
- major warp-stall sources.

---

## Numerical consistency follow-up

The initial implementation will focus on the native SM80 performance path.

A follow-up can investigate whether the native kernel can provide a fixed numerical execution topology for batch-invariant `linear_logp`.

Initial A100 characterization already shows that the current Triton fused path is bitwise invariant across the tested batch sizes and positions, while the materialized cuBLAS path can show small batch-shape-dependent differences.

A native SM80 implementation gives direct control over:

```text
GEMM accumulation order
+
vocabulary partitioning
+
online-softmax reduction
```

which makes it a useful basis for later deterministic / train-rollout alignment work.

Possible follow-up validation dimensions include:

- batch size;
- batch position;
- unrelated rows in the same batch;
- padding;
- fixed reduction topology.

This consistency work is separate from the initial native-kernel implementation.

---

## Out of scope for the initial implementation

The first implementation will not attempt to cover:

- SM90 changes;
- ROCm changes;
- tensor parallelism;
- distributed collectives;
- a new backward kernel;
- full train-rollout integration;
- FP8;
- architecture-specific tuning outside SM80.

The existing backward / recomputation path can remain unchanged initially.

---

## Expected contribution

This work would add a first-class native CUDA path for fused `linear_logp` on NVIDIA Ampere GPUs.

The contribution is intended to provide:

- better A100 `linear_logp` performance than the current portable path;
- the same key memory property of avoiding `[N, V]` logits materialization;
- an Ampere-specific CUDA implementation based on `cp.async` / Tensor Core MMA;
- a controllable execution path that can later be extended toward deterministic and distributed LogProb computation.

This complements the existing SM90 implementation and Triton fallback without changing their current behavior.

**Comment by Flink-ddd** [comment]
cc @KJLdefeated PTAL, is it necessary for us to support sm80 this architecture?

**Comment by QingYUDC** [comment]
> cc @KJLdefeated PTAL, is it necessary for us to support sm80 this architecture? 

Hi, I can open a PR for SM80 architecture support if required. It will take just a few hours.
